### ISSUE TITLE
Documenting and applying accessibility standards

### ISSUE THREAD

**Issue by micmath** [cferdinandi/kelp#13]
**Is your feature request related to a problem? Please describe.**
Finding a truly accessible UI library is impossible--I haven't found one that is 100% yet, and I feel like I've tried dozens. The _easiest_ time to start incorporating accessibility patterns into a UI library is at the beginning; it gets progressively harder with each added feature. KelpUI is in a fortunate position at the moment since it is still young and small. 

**Describe the solution you'd like**
I am proposing a standard that all new features will be expected to meet. The standard will consist of a list of accessibility testing scenarios that must be applied to any pull request and release: if the new feature does not meet this standard it will be rejected.

**Describe alternatives you've considered**
The alternative is to not be fully accessible or to leave it to hope and luck that your library doesn't exclude disabled users. I don't think this is a reasonable alternative but it wouldn't be a unique approach -- many UI libraries take this approach.

**Additional context**
I am volunteering to start and nudge along the discussion regarding what the accessibility standard for KelpUI should be.

For example, [WCAG 2.2 Level AA](https://www.gov.uk/service-manual/helping-people-to-use-your-service/understanding-wcag) seems like a reasonable base; exceptions could be made with justification after discussion.

Testingwise I would propose the following as a starting point:

1. **Screen Reader User, non-sighted**: Every feature should be fully understandable and usable with the screen turned off and a screen reader enabled (e.g. JAWS, NVDA, iOS VoiceOver, Android TalkBack).
2. **Keyboard Navigator**: Every feature should be fully understandable and usable with no mouse/touchpad cursor and either only a keyboard or a switch device available to interact with any component.
3. **No Sound**: Every feature should be fully understandable and usable with the volume set to zero on the user's device.
4. **No Color**: Every feature should be fully understandable and usable with the color mode set to monochrome on the user's device. [Available as a setting in Chrome DevTools: "Emulate achromatopsia (no color)"]
5. **Forced Color Mode**: Every feature should be fully understandable and usable with the device set to forced color mode (essentially black and white with no background images or colors). [Available as a setting in Chrome DevTools: "Emulate CSS forced-colors"]
6. **Contrast Reduced**: Every feature should be fully understandable and usable with the device set to reduced contrast. [Available as a setting in Chrome DevTools: "Emulate blurred vision" and "Emulate reduce contrast".]
7. **Screen Magnifier**: Every feature should be fully understandable and usable with page viewed through a software screen magnifier (where only a fraction of the screen is visible).
8. **Voice Control**: Every feature should be fully understandable and usable without touching the computer, using only voice control software.
9. **Images Turned Off**: Every feature should be fully understandable and usable with all images visually disabled (alternative text only).
10. **Alternative Navigation**: It should be possible to easily navigate the content of the page using only a list of headings, links, or HTML landmarks.

Other tests:

- HTML validation compliant
- Respects reduced motion preference settings -- follows guidance on flashing and blinking content
- Tabbing focus order and content source order matches the visual presentation
- Semantic elements are applied correctly in preference to ARIA fixes
- _probably more that I can't think of right now_ :-)

**Comment by cferdinandi** [comment]
I like this a a lot! And I’ve been building the library features with (most of) these in mind from the start. 

A lot of the “meat“ of this library is going to be in the interactive components, and I expect that a lot of those will have their own individual accessibility requirements around things like screen, reader, interactions, and keyboard navigation.

I’m not entirely sure how to standardize this yet, beyond adding an accessibility statement to the contributing guidelines.

I do plan to build out a section in the documentation around accessibility, as well as incorporating that into relevant sections elsewhere in the docs.

For example, the color palette is designed such that any two shades of a color that are at least 50 values apart meet WCAG AA requirements, and at 60 values apart, meet AAA requirements.

I would absolutely love your help in ensuring that this is as accessible as possible, with the understanding that accessibility isn’t some finish line you can cross because every disabled person has unique needs and challenges, and something that addresses the accessibility needs of one person can actually create issues for someone else.

I’ve got a huge to-do list around this project, so any help is greatly appreciated.

I don’t have a good idea of what the right solution is yet, and I’m very much open to suggestions.

**Comment by micmath** [comment]
If I were to come up with a list of the most useful first steps, in my opinion, they would be:

1. **Document your expectations** regarding accessibility and mention them on your pull request template. For example, "[x] Have you tested it using only the keyboard?" Just that _one check_ would avoid a whole class of barriers for disabled users.
2. This should include your project's **minimal supported browser/assistive technology combinations** for testing. [WebAIM Survey Results](https://webaim.org/projects/screenreadersurvey10/#used) are a good way to prioritize if you can only test a few combinations.
3. Recruit a few **accessibility champions** for the project, people who are experienced and willing to perform accessibility testing (actually-disabled assistive technology users would be preferred; setting up and learning to use JAWS, for example, is a full-time skill set all on its own). Encourage people to raise accessibility-related issues and treat them as you would functional bugs.
4. **Automated accessibility testing** will never get you 100% to your goal or, realistically, even to 50%, but it's a perfect way to knock out many of the low-hanging issues, so have an automated accessibility testing process documented and run it against all releases.

Doing just those four things, to whatever extent you can, would get you far along in being confident that your library could be used in organizations that are legally bound to conform to accessibility regulations, for example.

By the way, there are many free and open-source accessibility standards and policies you can use off-the-shelf or adapt to your own needs as you develop your project. I worked on [one for the BBC](https://www.bbc.co.uk/accessibility/forproducts/guides/html/), which is not 100% complete, but covers all the essential stuff.

**Comment by cferdinandi** [comment]
Dropping some resources in here for future reference:

- [Gov.uk](https://design-system.service.gov.uk/accessibility/)
- [React ARIA](https://react-spectrum.adobe.com/react-aria/accessibility.html)

**Comment by cferdinandi** [comment]
Kelp now has an accessibility statement with its goals, aims, and priorities clearly (I think) defined: https://kelpui.com/docs/getting-started/accessibility/

I also plan to add automated accessibility testing, using playwright, which is already handling component testing: https://playwright.dev/docs/accessibility-testing

Once that’s done, I think I can close this issue.

I don’t mean that to imply that “accessibility is done“ or anything like that. More that setting up some priorities and guard rails would be done, and then after that, it’s just and every day part of the development process.
